Sources/NIOCore/NIOTimer.swift

```diff
@@ -0,0 +1,94 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A type that handles timer callbacks scheduled with ``EventLoop/setTimer(for:_:)-5e37g``.
+///
+/// - Seealso: ``EventLoop/setTimer(for:_:)-5e37g``.
+public protocol NIOTimerHandler {
+    func timerFired(loop: any EventLoop)
+}
+
+/// An opaque handle that can be used to cancel a timer.
+///
+/// Users cannot create an instance of this type; it is returned by ``EventLoop/setTimer(for:_:)-5e37g``.
+///
+/// - Seealso: ``EventLoop/setTimer(for:_:)-5e37g``.
+public struct NIOTimer {
+    @usableFromInline
+    enum Backing {
+        /// A task created using `EventLoop.scheduleTask(deadline:_:)`, used by default for `EventLoop` implementations.
+        case scheduledTask(Scheduled<Void>)
+        /// An identifier for a timer, used by `EventLoop` implementations that conform to `CustomTimerImplementation`.
+        case custom(eventLoop: any NIOCustomTimerImplementation, id: UInt64)
+    }
+
+    @usableFromInline
+    var backing: Backing
+
+    fileprivate init(_ scheduled: Scheduled<Void>) {
+        self.backing = .scheduledTask(scheduled)
+    }
+
+    @inlinable
+    init(_ eventLoop: any NIOCustomTimerImplementation, id: UInt64) {
+        self.backing = .custom(eventLoop: eventLoop, id: id)
+    }
+
+    /// Cancel the timer associated with this handle.
+    @inlinable
+    public func cancel() {
+        switch self.backing {
+        case .scheduledTask(let scheduled):
+            scheduled.cancel()
+        case .custom(let eventLoop, let id):
+            eventLoop.cancelTimer(id)
+        }
+    }
+}
+
+/// Default implementation of `setSimpleTimer(for deadline:_:)`, backed by `EventLoop.scheduleTask`.
+extension EventLoop {
+    @discardableResult
+    public func setTimer(for deadline: NIODeadline, _ handler: any NIOTimerHandler) -> NIOTimer {
+        NIOTimer(self.scheduleTask(deadline: deadline) { handler.timerFired(loop: self) })
+    }
+}
+
+/// Default implementation of `setSimpleTimer(for duration:_:)`, delegating to `setSimpleTimer(for deadline:_:)`.
+extension EventLoop {
+    @discardableResult
+    @inlinable
+    public func setTimer(for duration: TimeAmount, _ handler: any NIOTimerHandler) -> NIOTimer {
+        self.setTimer(for: .now() + duration, handler)
+    }
+}
+
+/// Extension point for `EventLoop` implementations implement a custom timer.
+public protocol NIOCustomTimerImplementation: EventLoop {
```

Comment:
Is it necessary for us to add a new protocol? Any reason not to have `EventLoop` provide this directly?

Reply:
Yes, I tried to outline this in the PR description, but the goal here is to have a simple default implementation and have this as an opt-in extension point for implementations that want to do something different (e.g. more performant).

If we elevated these to the `EventLoop` protocol how would the default implementation look? Additionally, the protocol refinement here allows the type system to enforce that an implementation that's opting in here implements _both_ halves of the customisation point.

Reply:
I don't think we _ever_ want the default implementation: it's almost always worse than doing something bespoke. We have to provide it to avoid the API break, but we don't really care about it.

> If we elevated these to the EventLoop protocol how would the default implementation look?

More or less exactly as it does now. `cancel` is straightforward to implement, as you already have.

> Additionally, the protocol refinement here allows the type system to enforce that an implementation that's opting in here implements both halves of the customisation point.

This is handy, but doesn't necessarily outweigh the noise of the extra protocol.

More broadly, because we have added a new protocol, we make all event loops publicly offer these two methods:

```
    func setTimer(for deadline: NIODeadline, _ handler: any NIOTimerHandler) -> UInt64
    func cancelTimer(_ id: UInt64)
```

But we don't _want_ these two methods, we want the two that return typed handles. Nonetheless, users will see and be able to call these two on event loops. I think that's a strong enough API argument to reject this spelling.

Reply:
> More broadly, because we have added a new protocol, we make all event loops publicly offer these two methods:
> 
> ```
>     func setTimer(for deadline: NIODeadline, _ handler: any NIOTimerHandler) -> UInt64
>     func cancelTimer(_ id: UInt64)
> ```
> 
> But we don't _want_ these two methods, we want the two that return typed handles. Nonetheless, users will see and be able to call these two on event loops. I think that's a strong enough API argument to reject this spelling.

Presumably it would be only event loops that conform to this protocol that would publicly expose these methods and only when they are being held as a concrete type, rather than an `any EventLoop`?

However, I do concede that these symbols end up being public when ideally they should be internal.

I did this because I thought that was what the design called for: a slow default and the ability to provide something faster.

Maybe I've been thinking about exposing the wrong half of this. In the current PR I'm exposing the new methods (as a new protocol with new methods) and keeping the typed handle very opaque. Are you suggesting that I make the typed `NIOTimer` less opaque. Specifically that users can construct it as either a `scheduledTask` or a `custom` (will think on a better name if it's going to be public) which would allow event loop implementations to still opt in to provide a more performant implementation?

Reply:
OK, so I've gone ahead and made the patch that I believe is more in line with this feedback.

We no longer have another protocol, but instead we now have to be careful about ensuring we provide the right overloads for _both_ the schedule and cancel functions and need a runtime crash if the contract has been violated.

IMO, this was a place we could have had the type system help us and any external authors of event loops.

I appreciate the argument that it would result in the exposing of lower level functionality to end users, but they don't have to call that, and, if they did it would be safe still.

Possibly a moot point now, but would you have considered a layered approach like that initially proposed if we made use of `@_spi` to expose it only to event loop implementors?

The targeted relevant changes for this feedback are in 89c57ec15.

Reply:
> Possibly a moot point now, but would you have considered a layered approach like that initially proposed if we made use of `@_spi` to expose it only to event loop implementors?

I'm not convinced this would actually have worked. It's not immediately clear to me what the semantics of conformance to `@_spi` protocols actually _is_, so it's not clear to me that this would have prevented the protocol conformance being essentially public _or_ that it would have successfully populated the witness table.

If we assume that it would work as intended, though, I'm still not sure I'd have gone for it. I feel less strongly: the extra API noise is removed. But it would be a bit challenging to communicate to implementors how they should provide this implementation. Ultimately, we'd have to document it: otherwise it won't occur to implementors that they should implement this other protocol to implement these two functions. We have a lot of protocols, and it'll be easy to miss.

Once we're relying on documentation for correct implementation, I don't know that it's much worse to simply document that implementors must provide both methods. Yes, the type system will help correct implementations, but incorrect implementations will fail almost immediately the moment anyone tries to use them, including the implementor. So I'm just fundamentally not convinced that the downside risk is worth the extra abstractions.